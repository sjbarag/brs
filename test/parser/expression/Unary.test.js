const { Lexeme, BrsTypes, Parser } = require("brs");
const { Int32, BrsBoolean } = BrsTypes;
const BrsError = require("../../../lib/Error");

const { EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("unary expressions", () => {
        it("parses unary 'not'", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.True, text: "true", literal: BrsBoolean.True, line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary 'not'", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.Not, text: "not", line: 1 },
                { kind: Lexeme.True, text: "true", literal: BrsBoolean.True, line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses unary '-'", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Integer, text: "5", literal: new Int32(5), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses consecutive unary '-'", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Minus, text: "-", line: 1},
                { kind: Lexeme.Integer, text: "5", literal: new Int32(5), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});
