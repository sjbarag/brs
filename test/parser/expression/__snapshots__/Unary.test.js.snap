// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser unary expressions parses consecutive unary '-' 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Unary {
      "operator": Object {
        "kind": 7,
        "line": 1,
        "text": "-",
      },
      "right": Unary {
        "operator": Object {
          "kind": 7,
          "line": 1,
          "text": "-",
        },
        "right": Unary {
          "operator": Object {
            "kind": 7,
            "line": 1,
            "text": "-",
          },
          "right": Unary {
            "operator": Object {
              "kind": 7,
              "line": 1,
              "text": "-",
            },
            "right": Unary {
              "operator": Object {
                "kind": 7,
                "line": 1,
                "text": "-",
              },
              "right": Literal {
                "value": Int32 {
                  "kind": 3,
                  "value": 5,
                },
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`parser unary expressions parses consecutive unary 'not' 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Unary {
      "operator": Object {
        "kind": 67,
        "line": 1,
        "text": "not",
      },
      "right": Unary {
        "operator": Object {
          "kind": 67,
          "line": 1,
          "text": "not",
        },
        "right": Unary {
          "operator": Object {
            "kind": 67,
            "line": 1,
            "text": "not",
          },
          "right": Unary {
            "operator": Object {
              "kind": 67,
              "line": 1,
              "text": "not",
            },
            "right": Unary {
              "operator": Object {
                "kind": 67,
                "line": 1,
                "text": "not",
              },
              "right": Literal {
                "value": BrsBoolean {
                  "kind": 1,
                  "value": true,
                },
              },
            },
          },
        },
      },
    },
  },
]
`;

exports[`parser unary expressions parses unary '-' 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Unary {
      "operator": Object {
        "kind": 7,
        "line": 1,
        "text": "-",
      },
      "right": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 5,
        },
      },
    },
  },
]
`;

exports[`parser unary expressions parses unary 'not' 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Unary {
      "operator": Object {
        "kind": 67,
        "line": 1,
        "text": "not",
      },
      "right": Literal {
        "value": BrsBoolean {
          "kind": 1,
          "value": true,
        },
      },
    },
  },
]
`;
